# Puzzle 3

--- Day 3: Perfectly Spherical Houses in a Vacuum ---

Santa está entregando regalos a una cuadrícula bidimensional infinita de casas.

Comienza entregando un regalo a la casa en su lugar de inicio, y luego un elfo en el Polo Norte lo llama por radio y le dice dónde debe moverse a continuación. Los movimientos son siempre exactamente una casa al norte (^), sur (v), este (>) u oeste (<). Después de cada mudanza, entrega otro regalo a la casa en su nueva ubicación.

Sin embargo, el elfo en el polo norte ha tomado demasiado ponche de huevo, por lo que sus direcciones están un poco desviadas, y Santa termina visitando algunas casas más de una vez. ¿Cuántas casas reciben al menos un regalo?

Por ejemplo:

* ` > ` entrega regalos a 2 casas: una en el lugar de inicio y otra en el este.
* `^ > v <` entrega regalos a 4 casas en un cuadrado, incluso dos veces a la casa en su ubicación inicial / final.
* `^ v ^ v ^ v ^ v ^ v` entrega un montón de regalos a algunos niños muy afortunados en solo 2 casas.

## Logica de Resolucion

Como logica tome coordenadas bidimensionales, como la de los ejes cartesianos para el incio de santa que guardo en `posicion` y su valor son `[0, 0]` para los valores de `[x, y]` para tener cierta referencia, es decir comienza en el origen, luego cada uno de los caracteres dentro del archivo nos permitira saber como se va a desplazar Santa donde si por ejemplo:

* Partiendo desde el origen `[0, 0]`:
    * y el caracter que le toca es `^` su nueva posicion es `[0, 1]` entonces se habra dezplazado una posicion hacia arriba en el eje Y
    * y el caracter que le toca es `v` su nueva posicion es `[0, -1]` entonces se habra dezplazado una posicion hacia abajo en el eje Y
    * y el caracter que le toca es `>` su nueva posicion es `[1, 0]` entonces se habra dezplazado una posicion hacia la derecha en el eje X
    * y el caracter que le toca es `>` su nueva posicion es `[-1, 0]` entonces se habra dezplazado una posicion hacia la izquierda en el eje X

## Imagen para guiarse mejor
<img src='https://www.superprof.es/apuntes/wp-content/uploads/2018/06/eje-de-abscisas-y-coordenadas-cuadrantes.jpg-700x660.png' width='400px' height='400px'>

Luego esos valores vamos a añadirlos a una lista que llevara la cuenta del camino que va recorriendo Santa y que luego con un algoritmo de eliminacion de repetidos, nos quedaran solo los lugares, donde al menos haya pasado una vez

Al final mostramos el tamaño de esa lista y ese es nuestro resultado :)


## Mis Datos

Los datos que me dieron para el ejercicio si quieres practicar lo puedes encontrar en el archivo input_puzzle3.txt

## Codigo

In [11]:
contenido = open('input_puzzle3.txt', 'r')
mensaje = contenido.read()
camino = []
posicion = [0, 0]
camino.append(posicion)
for caracter in mensaje:
    if caracter == '^':
        movida = [posicion[0], posicion[1] + 1]
    if caracter == 'v':
        movida = [posicion[0], posicion[1] - 1]
    if caracter == '>':
        movida = [posicion[0] + 1, posicion[1]]
    if caracter == '<':
        movida = [posicion[0] - 1, posicion[1]]
    camino.append(movida)
    posicion = movida

result = []
for item in camino:
    if item not in result:
        result.append(item)

print(len(result))

2565


## Resultado Parte1: 2565

## Parte 2

--- La segunda parte ---
Al año siguiente, para acelerar el proceso, Santa crea una versión robótica de sí mismo, Robo-Santa, para entregarle regalos.

Santa y Robo-Santa comienzan en el mismo lugar (entregando dos regalos en la misma casa de inicio), luego se turnan para moverse según las instrucciones del elfo, que está leyendo con ponche de huevo el mismo guión del año anterior.

Este año, ¿cuántas casas reciben al menos un regalo?

Por ejemplo:

* `^ v` entrega regalos a 3 casas, porque Santa va al norte y luego Robo-Santa va al sur.
* `^ > v <` ahora entrega regalos a 3 casas, y Santa y Robo-Santa terminan donde empezaron.
* `^ v ^ v ^ v ^ v ^ v` ahora entrega regalos a 11 casas, con Santa yendo en una dirección y Robo-Santa en la otra.

## Logica de Resolucion


## Codigo

In [12]:
contenido = open('input_puzzle3.txt', 'r')
mensaje = contenido.read()
camino = []
posicion_santa = [0, 0]
posicion_robosanta = [0, 0]
camino.append(posicion_santa)
camino.append(posicion_robosanta)
turno = 1
movida_santa = posicion_santa
movida_robosanta = posicion_robosanta
for caracter in mensaje:
    if turno % 2 == 0:
        if caracter == '^':
            movida_santa = [posicion_santa[0], posicion_santa[1] + 1]
        if caracter == 'v':
            movida_santa = [posicion_santa[0], posicion_santa[1] - 1]
        if caracter == '>':
            movida_santa = [posicion_santa[0] + 1, posicion_santa[1]]
        if caracter == '<':
            movida_santa = [posicion_santa[0] - 1, posicion_santa[1]]
    else:
        if caracter == '^':
            movida_robosanta = [posicion_robosanta[0], posicion_robosanta[1] + 1]
        if caracter == 'v':
            movida_robosanta = [posicion_robosanta[0], posicion_robosanta[1] - 1]
        if caracter == '>':
            movida_robosanta = [posicion_robosanta[0] + 1, posicion_robosanta[1]]
        if caracter == '<':
            movida_robosanta = [posicion_robosanta[0] - 1, posicion_robosanta[1]]
    camino.append(movida_santa)
    camino.append(movida_robosanta)
    posicion_santa = movida_santa
    posicion_robosanta = movida_robosanta
    turno += 1
result = []
for item in camino:
    if item not in result:
        result.append(item)

print(len(result))

2639


## Resultado Parte2: 2639